File: ios/MoneyMove/NoticesSheet.swift
// Notes from whoever runs the game.
//
// One way only: the owner writes, players read, nobody replies. It exists for
// the two things a tournament needs a voice for — telling the field when the
// next round is, and telling a winner their prize is on its way — and being
// one-way is the point. A player who is sent a note cannot be dragged into a
// conversation they did not ask for.
//
// A note written to one person is marked as such. "You won" reads very
// differently from "everybody won", and the app should never blur the two.

import SwiftUI

struct Notice: Decodable, Equatable, Identifiable {
    var id: String
    var text: String
    var title: String?
    var at: Double
    /// Written to this player alone, rather than to the whole field.
    var personal: Bool = false
    var unread: Bool = false

    var date: Date { Date(timeIntervalSince1970: at / 1000) }
}

/// One poll for the tab: the bell needs a count whether or not the sheet is
/// open, so the count lives out here.
@MainActor
final class NoticeWatch: ObservableObject {
    @Published private(set) var notices: [Notice] = []
    @Published private(set) var unread = 0

    private weak var store: GameStore?
    private var task: Task<Void, Never>?

    func start(_ store: GameStore) {
        self.store = store
        guard task == nil else { return }
        task = Task { [weak self] in
            while !Task.isCancelled {
                await self?.load()
                try? await Task.sleep(for: .seconds(30))
            }
        }
    }

    func load() async {
        guard let store else { return }
        struct Feed: Decodable { var notices: [Notice] = []; var unread = 0 }
        guard let feed: Feed = try? await store.fetchJSON(
            "/api/notices?token=\(store.token)", raw: true) else { return }
        notices = feed.notices
        unread = feed.unread
    }

    /// Opening the list is reading it.
    func markRead() {
        guard unread > 0, let store else { return }
        unread = 0
        notices = notices.map { var n = $0; n.unread = false; return n }
        Task {
            struct Reply: Decodable { var ok: Bool? }
            let _: Reply? = try? await store.fetchJSON(
                "/api/notices/read", method: "POST", body: ["token": store.token])
        }
    }
}

/// The bell itself, for the top of the Social tab.
///
/// A control floating on the page, so it is a disc of glass rather than a
/// paper circle with a ring drawn round it. News is the brass bell and the
/// count; the gold outline the paper circle wore for it is gone, because the
/// rim is already brass and a second gold line beside it would be a sticker's
/// border. Nor does the glass take a tint for it — on 26 a tinted pane is a
/// slab of the colour, and a gold slab behind a gold bell hides the bell.
struct NoticeBell: View {
    @ObservedObject var watch: NoticeWatch
    var action: () -> Void

    @Environment(\.colorScheme) private var scheme
    @Environment(\.colorSchemeContrast) private var contrast
    @Environment(\.mmControlBackdrop) private var declared

    /// It stands beside the Social tab's title, on the page itself and not on
    /// a card, so that is its backdrop — unless whatever holds it has declared
    /// itself glass, in which case it is drawn on that glass instead of as a
    /// second pane over it.
    private var backdrop: BackdropKind { declared == .chrome ? .chrome : .page }

    var body: some View {
        let P = Palette.current(scheme)
        let news = watch.unread > 0
        Button(action: action) {
            Image(systemName: news ? "bell.badge.fill" : "bell.fill")
                .font(.system(size: 16, weight: .semibold))
                // Quiet, it is the glass's second ink: the third rank the
                // paper circle used measures 1.53:1 on glass.
                .foregroundStyle(news ? P.gold : backdrop.settledGlass(P)
                    .label(secondary: true, increaseContrast: contrast == .increased))
                .frame(width: 38, height: 38)
        }
        .buttonStyle(MMButtonStyle(kind: .ghost, form: .disc))
        .mmControls(on: backdrop)
        .accessibilityLabel(news ? "\(watch.unread) unread notes" : "Notes")
        // The count rides outside the glass, over its shoulder: the material
        // is free to clip what it holds to its own circle, and a badge is
        // meant to break the edge.
        .overlay(alignment: .topTrailing) {
            if news {
                Text("\(min(watch.unread, 9))")
                    .font(.system(size: 10, weight: .black, design: .rounded))
                    .foregroundStyle(P.accentInk)
                    .frame(width: 17, height: 17)
                    .background(P.red, in: Circle())
                    .offset(x: 3, y: -2)
                    .allowsHitTesting(false)
                    .accessibilityHidden(true)
            }
        }
    }
}

struct NoticesSheet: View {
    @ObservedObject var watch: NoticeWatch

    @Environment(\.dismiss) private var dismiss
    @Environment(\.colorScheme) private var scheme

    var body: some View {
        let P = Palette.current(scheme)
        NavigationStack {
            ScrollView {
                if watch.notices.isEmpty {
                    VStack(spacing: 9) {
                        Image(systemName: "bell.slash")
                            .font(.system(size: 26))
                            .foregroundStyle(P.ink3)
                        Text("Nothing yet")
                            .font(.system(size: 15, weight: .heavy, design: .rounded))
                            .foregroundStyle(P.ink2)
                        Text("Announcements about tournaments — when a round opens, when a prize is on its way — turn up here.")
                            .font(.system(size: 12.5, weight: .medium, design: .rounded))
                            .foregroundStyle(P.ink3)
                            .multilineTextAlignment(.center)
                            .fixedSize(horizontal: false, vertical: true)
                    }
                    .frame(maxWidth: .infinity)
                    .padding(.horizontal, 30)
                    .padding(.top, 70)
                } else {
                    VStack(spacing: 10) {
                        ForEach(watch.notices) { n in row(n, P) }
                    }
                    .padding(16)
                }
            }
            .navigationTitle("Notes")
            .navigationBarTitleDisplayMode(.inline)
            .toolbar {
                ToolbarItem(placement: .topBarTrailing) {
                    Button("Done") { dismiss() }
                        .sheetBarItem(on: .page)
                }
            }
        }
        .mmControls(on: .platter(.page))
        .sheetPaper(P.page)
        .task {
            await watch.load()
            // Reading it is reading it — but leave the dots up for a beat so
            // somebody can see which ones were new.
            try? await Task.sleep(for: .seconds(1.2))
            watch.markRead()
        }
    }

    private func row(_ n: Notice, _ P: Palette) -> some View {
        VStack(alignment: .leading, spacing: 6) {
            HStack(spacing: 7) {
                if n.personal {
                    Text("FOR YOU")
                        .font(.system(size: 8.5, weight: .black, design: .rounded))
                        .kerning(0.6)
                        .foregroundStyle(P.accentInk)
                        .padding(.vertical, 2).padding(.horizontal, 7)
                        .background(P.gold, in: Capsule())
                }
                if let t = n.title, !t.isEmpty {
                    Text(t)
                        .font(.system(size: 14.5, weight: .heavy, design: .rounded))
                        .foregroundStyle(P.ink)
                }
                Spacer(minLength: 4)
                if n.unread { Circle().fill(P.red).frame(width: 8, height: 8) }
            }
            Text(n.text)
                .font(.system(size: 13.5, weight: .medium, design: .rounded))
                .foregroundStyle(P.ink2)
                .fixedSize(horizontal: false, vertical: true)
            Text(n.date.formatted(.dateTime.weekday(.abbreviated).day().month(.abbreviated).hour().minute()))
                .font(.system(size: 11, weight: .semibold, design: .rounded))
                .foregroundStyle(P.ink3)
        }
        .frame(maxWidth: .infinity, alignment: .leading)
        .padding(13)
        .background(n.personal ? P.goldSoft : P.sunken,
                    in: RoundedRectangle(cornerRadius: 14, style: .continuous))
        .overlay(RoundedRectangle(cornerRadius: 14, style: .continuous)
            .stroke(n.personal ? P.gold : P.rule, lineWidth: 1))
    }
}

File: android/app/src/main/java/com/moneymove/game/NoticesSheet.kt
package com.moneymove.game

import android.text.format.DateFormat
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxHeight
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.offset
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.lazy.rememberLazyListState
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.Text
import androidx.compose.material3.rememberModalBottomSheetState
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch

/**
 * Notes from whoever runs the game.
 *
 * One way only: the owner writes, players read, nobody replies. It is here for
 * the two things a tournament needs a voice for — telling the field when the
 * next round opens, and telling a winner their prize is on its way — and being
 * one-way is the point, because somebody who is sent a note cannot be dragged
 * into a conversation they never asked for.
 *
 * A note written to one player is marked as such. "You won" reads very
 * differently from "everybody won", and the list must never blur the two.
 *
 * The bell and the list are separate composables on purpose: the count has to
 * be visible from a top bar long before anybody taps it, so it comes from
 * [MessagingStore]'s own poll rather than from this sheet having been opened.
 */

/**
 * The bell, for the top of the Social tab — iOS's NoticeBell. [count] is
 * `messaging.unread`.
 *
 * iOS draws its own bell here, the filled one, and the bell with a dot on it
 * once something is waiting; icons.js has no bell, so SafetySheets.kt draws
 * both on the glyph grid. Quiet ink with nothing waiting, gold with
 * something, and the count in a red coin off the top corner that stops at
 * nine, as iOS's does.
 *
 * No sound of its own: iOS's bell knocks through whoever hosts it, and the
 * Social tab does that.
 */
@Composable
fun NoticeBell(count: Int, modifier: Modifier = Modifier, onClick: () -> Unit) {
    val p = P.current
    val waiting = count > 0
    Box(
        modifier
            .size(38.dp)
            .semantics { contentDescription = if (waiting) "$count unread notes" else "Notes" },
    ) {
        Box(
            Modifier
                .size(38.dp)
                .clip(CircleShape)
                .background(p.sunken)
                .border(1.dp, if (waiting) p.gold else p.rule, CircleShape)
                .clickable { onClick() },
            contentAlignment = Alignment.Center,
        ) {
            SfMark(if (waiting) "bell.badge.fill" else "bell.fill", 20.dp, if (waiting) p.gold else p.ink3)
        }
        if (waiting) {
            Box(
                Modifier
                    .align(Alignment.TopEnd)
                    .offset(x = 3.dp, y = (-2).dp)
                    .size(17.dp)
                    .clip(CircleShape)
                    .background(p.red),
                contentAlignment = Alignment.Center,
            ) {
                Text(
                    "${count.coerceAtMost(9)}",
                    color = p.accentInk, fontSize = 10.sp, fontWeight = FontWeight.Black,
                )
            }
        }
    }
}

/**
 * The list — iOS's NoticesSheet: the full height of the screen on the
 * sheet's glass ([MMSheet]; iOS painted this one the page colour, and on
 * iOS 26 it is the system's glass too), "Notes" in the middle of the bar and
 * Done on the right, a card per note, and a bell with a line through it when
 * there are none.
 */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun NoticesSheet(messaging: MessagingStore, onDismiss: () -> Unit) {
    val p = P.current
    val sheetState = rememberModalBottomSheetState(skipPartiallyExpanded = true)
    val scope = rememberCoroutineScope()
    val notices = messaging.notices

    fun close() {
        scope.launch { sheetState.hide() }.invokeOnCompletion { onDismiss() }
    }

    // Reading it is reading it — but the dots stay up for a beat so somebody
    // can see which ones were new, then go, as iOS's do. The server is told
    // at the same moment; until the next poll brings the notes back unmarked
    // it is this that keeps the dots down.
    var read by remember { mutableStateOf(false) }
    LaunchedEffect(Unit) {
        // Waited on rather than fired and forgotten: marking a list read
        // before it has arrived says "seen" about notes nobody has been shown
        // yet, and the server stamps the moment it is told rather than
        // anything this app sends, so there is no taking it back.
        messaging.refreshNotices().join()
        delay(1_200)
        messaging.markNoticesRead()
        read = true
    }

    MMSheet(
        onDismissRequest = onDismiss,
        sheetState = sheetState,
    ) {
        Column(Modifier.fillMaxWidth().fillMaxHeight()) {
            NotesBar { close() }
            if (notices.isEmpty()) {
                Column(
                    Modifier
                        .fillMaxWidth()
                        .weight(1f)
                        .verticalScroll(rememberScrollState())
                        .padding(horizontal = 30.dp)
                        .padding(top = 70.dp),
                    horizontalAlignment = Alignment.CenterHorizontally,
                    verticalArrangement = Arrangement.spacedBy(9.dp),
                ) {
                    // The empty state stands straight on the sheet's glass,
                    // so it takes the glass's quiet ink rather than ink3.
                    SfMark("bell.slash", 32.dp, quietInk())
                    Text(
                        "Nothing yet",
                        color = p.ink2, fontSize = 15.sp, fontWeight = FontWeight.ExtraBold,
                    )
                    Text(
                        "Announcements about tournaments — when a round opens, when a prize " +
                            "is on its way — turn up here.",
                        color = quietInk(), fontSize = 12.5.sp, lineHeight = 17.sp,
                        fontWeight = FontWeight.Medium, textAlign = TextAlign.Center,
                    )
                }
            } else {
                val list = rememberLazyListState()
                LazyColumn(
                    state = list,
                    // The notes fade out under the bar rather than slicing off.
                    modifier = Modifier.fillMaxWidth().weight(1f).scrollEdge(list),
                    contentPadding = PaddingValues(16.dp),
                    verticalArrangement = Arrangement.spacedBy(10.dp),
                ) {
                    items(notices) { n ->
                        NoticeRow(n, isNew = n.unread && !read)
                    }
                }
            }
        }
    }
}

/**
 * iOS's inline bar: "Notes" in the middle, Done on the right in the accent —
 * regular weight, as iOS's trailing toolbar button is.
 */
@Composable
private fun NotesBar(onDone: () -> Unit) {
    val p = P.current
    Box(Modifier.fillMaxWidth().height(44.dp).padding(horizontal = 16.dp)) {
        Text(
            "Notes",
            color = p.ink, fontSize = 17.sp, fontWeight = FontWeight.SemiBold,
            modifier = Modifier.align(Alignment.Center),
        )
        Text(
            "Done",
            color = p.red, fontSize = 17.sp, fontWeight = FontWeight.Normal,
            modifier = Modifier
                .align(Alignment.CenterEnd)
                .clip(RoundedCornerShape(8.dp))
                .clickable { onDone() }
                .padding(horizontal = 4.dp, vertical = 6.dp),
        )
    }
}

@Composable
private fun NoticeRow(n: Notice, isNew: Boolean) {
    val p = P.current
    val shape = RoundedCornerShape(14.dp)
    Column(
        Modifier
            .fillMaxWidth()
            .clip(shape)
            // Gold for a note written to this player alone. Whoever reads it
            // should be able to tell, at a glance and from across the room,
            // that it is about them.
            .background(if (n.personal) p.goldSoft else p.sunken)
            .border(1.dp, if (n.personal) p.gold else p.rule, shape)
            .padding(13.dp),
        verticalArrangement = Arrangement.spacedBy(6.dp),
    ) {
        // The heading line is always there, as iOS's is, even when it holds
        // nothing but the space before the words.
        Row(verticalAlignment = Alignment.CenterVertically) {
            if (n.personal) {
                Text(
                    "FOR YOU",
                    color = p.accentInk,
                    fontSize = 8.5.sp,
                    letterSpacing = 0.6.sp,
                    fontWeight = FontWeight.Black,
                    modifier = Modifier
                        .clip(RoundedCornerShape(99.dp))
                        .background(p.gold)
                        .padding(horizontal = 7.dp, vertical = 2.dp),
                )
                Spacer(Modifier.width(7.dp))
            }
            if (n.title.isNotBlank()) {
                Text(
                    n.title,
                    color = p.ink, fontSize = 14.5.sp, fontWeight = FontWeight.ExtraBold,
                    modifier = Modifier.weight(1f),
                )
            } else {
                Spacer(Modifier.weight(1f))
            }
            Spacer(Modifier.width(4.dp))
            if (isNew) {
                Box(Modifier.size(8.dp).clip(CircleShape).background(p.red))
            }
        }
        Text(
            n.text,
            color = p.ink2, fontSize = 13.5.sp, lineHeight = 19.sp,
            fontWeight = FontWeight.Medium,
        )
        Text(
            whenWritten(n.at),
            color = p.ink3, fontSize = 11.sp, fontWeight = FontWeight.SemiBold,
        )
    }
}

/**
 * When it was written, the way iOS writes it: the short weekday, the day and
 * the short month, and the time — "Thu, 25 Sep, 14:32" in the reader's own
 * locale's order, which is what `.dateTime` with those fields asks of the
 * phone. The clock is the one the phone's 24-hour switch picks, not the
 * locale's habit, as on iOS — the same rule the cup's times follow (see
 * [clockText]).
 */
@Composable
private fun whenWritten(at: Double): String {
    val twentyFour = DateFormat.is24HourFormat(LocalContext.current)
    if (at.toLong() <= 0L) return ""
    return clockText(at, "EEEdMMMjmm", twentyFour)
}
